### PR TITLE
Optionally bootstrap 1-3 monitoring nodes

### PR DESCRIPTION
**Pull request by kelvk**
As head nodes do not run monitoring, this updates the bootstrap process to optionally deploy additional VM(s) for the monitoring role. Monitoring bootstrapping is disabled by default, as anecdotally, a functional Openstack cluster is generally the requirement.

### CODE REVIEW

**Comment by erhudy** [comment]
I shall give this a test. One minor thought before I start: right now this uses the configure block for all cluster nodes, which allocates 4 extra volumes for Ceph per VM. These aren't used by the monitoring role, right?

The reason I ask is I have a to-do item to consolidate the two configure blocks into one down the road since there's a lot of duplicated material between the two and I'd like to revise it to be a single configuration block with appropriate guards depending on the VM's role, but this quickly becomes a more extensive refactoring of how the Vagrant bootstrap process creates VMs and determines their role, and implies that we would want to bring in something like `cluster.txt` to handle this whole process in a manner consistent with hardware clusters.

**Comment by kelvk** [comment]
Correct. The monitoring role uses just the first device. I agree we would need to move on to a cluster definition approach as roles are increasingly fragmented. There had been occasions when I did not need work nodes. In the meantime, this PR hopefully will make testing monitoring more convenient.

**Comment by erhudy** [comment]
I agree, I think that as long as everything works here this can be merged as-is and then we'll come back later to try to make it all consistent.

File: bootstrap/vagrant_scripts/vagrant_configure_chef.sh [hunk 1]
```diff
@@ -81,29 +80,61 @@ do_on_node bootstrap "$KNIFE cookbook upload apt bcpc chef-client cron logrotate
...
   # run on head node one last time to update HAProxy with work node IPs
   do_on_node vm1 "sudo chef-client"
   # HUP OpenStack services on each node to ensure everything's in a working state
-  for vm in vm1 vm2 vm3; do
+  for vm in $vms; do
     do_on_node $vm "sudo hup_openstack || true"
   done
+  # Run chef on each mon VM before assigning next node for monitoring.
+  for vm in $mon_vms; do
+    do_on_node bootstrap "$KNIFE node run_list set bcpc-$vm.$BOOTSTRAP_DOMAIN 'role[BCPC-Monitoring]'"
```
Comment (erhudy) [inline bootstrap/vagrant_scripts/vagrant_configure_chef.sh]: Can you migrate this section up to where the other roles are set? My initial reasoning behind the `BOOTSTRAP_CHEF_DO_CONVERGE` variable was that it should stop just before converge, and all the roles/environments should be assigned out and ready to go.

**Comment by erhudy** [comment]
So, I tested with 3 nodes and 0 nodes and it works as expected in both cases.

One problem I have found that is really a Vagrant issue (you can argue whether it's a bug or not), is that Vagrant reads the Vagrantfile every time you run a command instead of referencing metadata in the `.vagrant` folder, should the folder exist. What this means is that Vagrant effectively loses track of the extra monitoring VMs unless `MONITORING_NODES` is set every time you run a Vagrant command. Observe:
```
➜  vagrant_scripts git:(09922a9) vagrant status          
Current machine states:

bootstrap                 not created (virtualbox)
vm1                       not created (virtualbox)
vm2                       not created (virtualbox)
vm3                       not created (virtualbox)

This environment represents multiple VMs. The VMs are all listed
above with their current state. For more information about a specific
VM, run `vagrant status NAME`.
➜  vagrant_scripts git:(09922a9) MONITORING_NODES=3 vagrant status
Current machine states:

bootstrap                 not created (virtualbox)
vm1                       not created (virtualbox)
vm2                       not created (virtualbox)
vm3                       not created (virtualbox)
vm4                       not created (virtualbox)
vm5                       not created (virtualbox)
vm6                       not created (virtualbox)

This environment represents multiple VMs. The VMs are all listed
above with their current state. For more information about a specific
VM, run `vagrant status NAME`.
```
Basically what this means is that Vagrant will lose track of the extra monitoring VMs unless that environment variable is set, which can lead to some weirdly counterintuitive behavior (Vagrant not destroying all VMs if you reduce the number of monitoring nodes between cluster rebuilds, for example, or not being able to `vagrant ssh` to VMs that you know are running).

File: bootstrap/vagrant_scripts/vagrant_configure_chef.sh (hunk 1 above)
Comment (kelvk) [inline bootstrap/vagrant_scripts/vagrant_configure_chef.sh]: Probably, but likely not without reworking some recipes. I attempted to and found out that uninitialized node['bcpc']['management']['ip'] of second/third nodes would break during execution of haproxy/mysql/etc recipes. We used to knife bootstrap with role, so this hadn't been an issue until now.

**Comment by erhudy** [comment]
The above comment wasn't meant to be a reason why this shouldn't be merged, just an observation/caution about counterintuitive Vagrant behavior that will be encountered after the merge if you start adding monitoring nodes.

File: bootstrap/vagrant_scripts/vagrant_configure_chef.sh (hunk 1 above)
Comment (erhudy) [inline bootstrap/vagrant_scripts/vagrant_configure_chef.sh]: That section only sets the role on the Chef server, it doesn't perform any convergence (it's so that you can go to the node and just run `chef-client`). If somebody runs it out of order, it's their own fault.

**Comment by kelvk** [comment]
Yes, that's rather annoying and I think the presence of a cluster definition file, if it is then used to generate the Vagrant file (thinking aloud here), could lay this to rest.

**Comment by kelvk** [comment]
Added a note about this behavior in `FAQ.md`. `vagrant_clean.sh` is also updated to ensure mon VMs are always destroyed.

File: bootstrap/vagrant_scripts/vagrant_configure_chef.sh (hunk 1 above)
Comment (erhudy) [inline bootstrap/vagrant_scripts/vagrant_configure_chef.sh]: Talked to @kelvk, rescinding this request.

**Comment by erhudy** [comment]
Built fine with 3 monitoring nodes, doing one last build with no monitoring nodes and then will merge if successful.

**Comment by erhudy** [comment]
Opened #648 to track the issue discussed above, merging.
